_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+This overhead includes the following:
+
+- A small number of Remote Procedure Calls (RPCs) to register the partitions with the coordinator. 
+  These RPCs can be batched together to reduce overhead.
+- A marker that must be written to each partition when the transaction is completed.
+- A few state change records that are appended to the internal transaction state log.
+
+Most of this latency is on the producer side, where the transactions API is implemented.
+The performance degradation can be significant when having short transaction commit intervals.
+Increasing the transaction duration also increases the end-to-end latency and may require some additional tuning, so it's a matter of tradeoff.
+
+Another potential drawback is hanging transactions, which are transactions with a missing or out-of-order control record.
+They may be caused by a network issue causing delayed messages ([KIP-890](https://cwiki.apache.org/confluence/display/KAFKA/KIP-890%3A+Transactions+Server-Side+Defense)).
+This issue is rare, but it's important to be aware and set alerts, because it can impact service availability.
+
+A hanging transaction is usually revealed by a stuck application with growing lag ([KIP-664](https://cwiki.apache.org/confluence/display/KAFKA/KIP-664%3A+Provide+tooling+to+detect+and+abort+hanging+transactions)).
+Despite messages being produced, consumers can't make any progress because the LSO does not grow anymore.
+
+For example, you will see the following messages in the consumer logs:
+
+```sh
+[Consumer clientId=my-client, groupId=my-group] The following partitions still have unstable offsets which are not cleared on the broker side: [__consumer_offsets-27], 
+this could be either transactional offsets waiting for completion, or normal offsets waiting for replication after appending to local log
+```
+
+Additionally, if the partition belongs to a compacted topic, this causes the unbounded partition growth, because the `LogCleaner` does not clean beyond the LSO.
+If not detected and fixed in time, this may lead to disk space exhaustion and service disruption.
+
+Fortunately, there are embedded tools that can be used to find all hanging transactions and roll them back.
+When doing that, it is important to understand the business context surrounding the records in this transaction and possibly append those records again using a new transaction.
+
+```sh
+# find all hanging transactions in a broker
+$ $KAFKA_HOME/bin/kafka-transactions.sh --bootstrap-server :9092 find-hanging --broker 0
+Topic                  Partition   ProducerId  ProducerEpoch   StartOffset LastTimestamp               Duration(s)
+__consumer_offsets     27          171100      1               913095344   2022-06-06T03:16:47Z        209793
+
+# abort an hanging transaction
+$ $KAFKA_HOME/bin/kafka-transactions.sh --bootstrap-server :9092 abort --topic __consumer_offsets --partition 27 --start-offset 913095344
+```
+
+### Basic example
```

Comment:
I would prefer to keep it last. I don't think you can really predict how a reader will consume this content, because it mostly depends on prior knowledge. I do expect some back and forth between theory and example. The impatient reader may want to jump straight into the example and then come back to the theory.

Reply:
I am not sure to agree ... or maybe this article is for people who already have a good understand of transactions already.